rust/src/push/mod.rs

```diff
@@ -0,0 +1,411 @@
+use serde_json::Value;
+
+mod base_rules;
+
+/// Called when registering modules with python.
+pub fn register_module(py: Python<'_>, m: &PyModule) -> PyResult<()> {
+    let child_module = PyModule::new(py, "push")?;
+    child_module.add_class::<PushRule>()?;
+    child_module.add_class::<PushRules>()?;
+    child_module.add_class::<FilteredPushRules>()?;
+    m.add_submodule(child_module)?;
+
+    // We need to manually add the module to sys.modules to make `from
+    // synapse.synapse_rust import push` work.
+    py.import("sys")?
+        .getattr("modules")?
+        .set_item("synapse.synapse_rust.push", child_module)?;
+
+    Ok(())
+}
+
+/// A single push rule for a user.
+#[derive(Debug, Clone)]
+#[pyclass(frozen)]
+pub struct PushRule {
+    pub rule_id: Cow<'static, str>,
+    #[pyo3(get)]
+    pub priority_class: i32,
+    pub conditions: Cow<'static, [Condition]>,
+    pub actions: Cow<'static, [Action]>,
+    #[pyo3(get)]
+    pub default: bool,
+    #[pyo3(get)]
+    pub default_enabled: bool,
+}
+
+#[pymethods]
+impl PushRule {
+    #[staticmethod]
+    pub fn from_db(
+        rule_id: String,
+        priority_class: i32,
+        conditions: &str,
+        actions: &str,
+    ) -> Result<PushRule, Error> {
+        let conditions = serde_json::from_str(conditions).context("parsing conditions")?;
+        let actions = serde_json::from_str(actions).context("parsing actions")?;
+
+        Ok(PushRule {
+            rule_id: Cow::Owned(rule_id),
+            priority_class,
+            conditions,
+            actions,
+            default: false,
+            default_enabled: true,
+        })
+    }
+
+    #[getter]
+    fn rule_id(&self) -> &str {
+        &self.rule_id
+    }
+
+    #[getter]
+    fn actions(&self) -> Vec<Action> {
+        self.actions.clone().into_owned()
+    }
+
+    #[getter]
+    fn conditions(&self) -> Vec<Condition> {
+        self.conditions.clone().into_owned()
+    }
+
+    fn __repr__(&self) -> String {
+        format!(
+            "<PushRule rule_id={}, conditions={:?}, actions={:?}>",
+            self.rule_id, self.conditions, self.actions
+        )
+    }
+}
+
+/// The "action" Synapse should perform for a matching push rule.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Action {
+    DontNotify,
+    Notify,
+    Coalesce,
+    SetTweak(SetTweak),
+}
+
+impl IntoPy<PyObject> for Action {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        pythonize(py, &self).expect("valid action")
+    }
+}
+
+/// The body of a `SetTweak` push action.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub struct SetTweak {
+    set_tweak: Cow<'static, str>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    value: Option<TweakValue>,
+
+    // This picks saves any other fields that may have been added as clients.
+    // These get added when we convert the `Action` to a python object.
+    #[serde(flatten)]
+    other_keys: Value,
```

Comment:
Yeah, `Value` is an enum that represents any arbitrary JSON value: https://docs.rs/serde_json/latest/serde_json/value/enum.Value.html

Reply:
From Rust's point of view I'd expect that `SetTweak.other_keys` is a mapping of some kind from strings to Values; I'm not quite sure what `other_keys: Value` means. Though I haven't bothered to RTFM for `serde(flatten)` which might be enlightening...

Reply:
I don't think one can say `other_keys: Value::Object` though because `Value::Object` isn't a proper type?

Reply:
> From Rust's point of view I'd expect that `SetTweak.other_keys` is a mapping of some kind from strings to Values; I'm not quite sure what `other_keys: Value` means. Though I haven't bothered to RTFM for `serde(flatten)` which might be enlightening...

The way [`#[serde(flatten)]`](https://serde.rs/attr-flatten.html) works is basically to take any remaining fields left over from the deserializing into the outer struct, and trying to deserialise them into the inner struct.

`other_keys` in this instance would *always* be `Value::Object` (since we're deserializing a JSON object), so we could absolutely replace it with `Map<String, Value>` instead, but that just involves a little bit more typing on my part.